design/one-pager-default-resource-class.md

```diff
@@ -88,66 +118,241 @@ spec:
+1. Denote the default policy via label.
+
+### Multiple Policies Per Namespace
+
+Currently, default class controllers, which act on claims that omit a
+`classRef`, will fail to reconcile if multiple `Policies` exist within the
+claim's namespace. For example, if two `MySQLInstancePolicies` are created in
+namespace `crossplane-system`, then a `MySQLInstance` claim created in
+`crossplane-system` that omits a `classRef` will not be assigned either one of
+the classes referenced by the two `MySQLInstancePolicy` objects respectively.
+
+In order to reintroduce the ability to select a class of service in claim, the
+following steps can be taken:
+
+1. Update all `Policy` types to include a `classRef` field instead of a
+   `defaultClassRef` field. This would require updating the `crossplane-runtime`
+   embeddable struct
+   [`Policy`](https://github.com/crossplaneio/crossplane-runtime/blob/e4d61ee2805af680baf16fc2a1d8f79538d0f9bb/apis/core/v1alpha1/resource.go#L93),
+   then bumping the dependency in core Crossplane.
+1. Remove the `classRef` field from all claims entirely. This would require the
+   same steps as above but with
+   [`ResourceClaimSpec`](https://github.com/crossplaneio/crossplane-runtime/blob/e4d61ee2805af680baf16fc2a1d8f79538d0f9bb/apis/core/v1alpha1/resource.go#L49).
+1. Add a `policyRef` field to all claims using the same method detailed above.
+   This would require updating the same steps as above also with
+   [`ResourceClaimSpec`](https://github.com/crossplaneio/crossplane-runtime/blob/e4d61ee2805af680baf16fc2a1d8f79538d0f9bb/apis/core/v1alpha1/resource.go#L49).
+1. Alter all claim controllers (which live in provider stacks) to only reconcile
+   claims that have a `policyRef` to a `Policy` that has a `classRef` of that
+   kind's strongly-typed class. For instance, the `RDSInstance` claim controller
+   will only reconcile `MySQLInstance` claims that have `policyRef` to a
+   `MySQLInstancePolicy` that has a `classRef` to an `RDSInstanceClass`. This
+   will involve updating the claim controller
+   [predicates](https://github.com/crossplaneio/crossplane-runtime/blob/e4d61ee2805af680baf16fc2a1d8f79538d0f9bb/pkg/resource/predicates.go#L41),
+   as well as the actual logic of the [shared claim
+   reconciler](https://github.com/crossplaneio/crossplane-runtime/blob/e4d61ee2805af680baf16fc2a1d8f79538d0f9bb/pkg/resource/claim_reconciler.go#L281),
+   both in `crossplane-runtime`. It should require minimal updates to the actual
+   claim controllers in each of the provider stacks in order to indicate the
+   policy `kind` that they should use.
+
+In this model, a claim will now look as follows:
 
 ```yaml
-apiVersion: storage.crossplane.io/v1alpha1
-kind: NoSQLInstance
+apiVersion: database.crossplane.io/v1alpha1
+kind: MySQLInstance
 metadata:
-  name: dynamo-claim
-  namespace: demo
+  name: mysql-claim
 spec:
-  ...
+  policyRef:
+    name: standard-mysql
+    namespace: crossplane-system
```

Comment:
currently in master, policies need to live in the same namespace as the claim.  Is this a new ability in this PR as well, that policies can live in a different namespace from a claim that references them?

Reply:
Now that policies can be referenced by the claim, they can reference a policy that exists in a different namespace (just like they can currently reference a class in a different namespace). If a `policyRef` was omitted, they would only be able to fall back on the policy specified as `default` in their namespace though.

Reply:
I think this thread might be out of date now that we're planning to use portable classes rather than policies.

Given defaulting behaviour will be namespace scoped (i.e. we'll look for a portable class with label `default="true"` in the _namespace of the claim_) I suggest we consider requiring that portable classes live in the same namespace as the claims that reference them; i.e. we use a [`corev1.LocalObjectReference`](https://godoc.org/k8s.io/api/core/v1#LocalObjectReference) from claim to (portable) class.

As long as we use a [`corev1.ObjectReference`](https://godoc.org/k8s.io/api/core/v1#ObjectReference) from the portable class to its underlying non-portable class this still gives us flexibility on which namespace the non-portable class (and thus managed resource) are created in.

Consider the following flow:

1. Infrastructure operator creates infra namespace (e.g. `prod`).
1. Infrastructure operator creates a provider and some non-portable classes (e.g. `RDSInstanceClass`) in said namespace.
1. Infrastructure operator "publishes" the aforementioned non-portable classes to the app operator's namespace by creating portable some classes (e.g. `MySQLInstanceClass`) that reference the non-portable classes created in step 2, and labelling one of each kind as the default for said app namespace.
1. App operator lists the portable classes that are available in their app namespace, picks one, and creates a resource claim that references it.